___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright by Pierian Data Inc.</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# DataFrames

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [66]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [67]:
# help(pd.DataFrame)

Generate a NumPy array from the random seed set 101, with the dimesnions (4,3).
Make sure the seed is in the same cell as the random call -
- https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do

In [68]:
np.random.seed(101)
mydata = np.random.randint(0,101,(4,3))

In [69]:
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

Create an index for the data -

In [70]:
myindex = ['CA','NY','AZ','TX']

List for the column headings -

In [71]:
mycolumns = ['Jan','Feb','Mar']

Create the dataframe using the 'DataFrame' function, passing in only the data (not the indexes or coumn headings) -

In [72]:
df = pd.DataFrame(data=mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

Now pass in the index values -

In [73]:
df = pd.DataFrame(data=mydata,index=myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

And finally the columns -

In [74]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

Reports back information for this dataframe....

In [75]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Jan     4 non-null      int32
 1   Feb     4 non-null      int32
 2   Mar     4 non-null      int32
dtypes: int32(3)
memory usage: 80.0+ bytes



# Reading a .csv file for a DataFrame

----

## NOTE: We will go over all kinds of data inputs and outputs (.html, .csv, .xlxs , etc...) later on in the course! For now we just need to read in a simple .csv file.

----

## CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [76]:
pwd

'C:\\Users\\aaron\\Documents\\Programming\\Python\\Machine Learning\\Masterclass Notes\\03-Pandas'

#### List the files in your current directory with 'ls'

In [77]:
ls

 Volume in drive C is OS
 Volume Serial Number is 6231-9E9F

 Directory of C:\Users\aaron\Documents\Programming\Python\Machine Learning\Masterclass Notes\03-Pandas

14/05/2021  09:56 AM    <DIR>          .
14/05/2021  09:56 AM    <DIR>          ..
08/05/2021  09:59 AM    <DIR>          .ipynb_checkpoints
08/05/2021  09:59 AM    <DIR>          __pycache__
12/05/2021  03:55 PM            14,202 00-Series.ipynb
14/05/2021  09:56 AM           156,757 01-DataFrames.ipynb
08/05/2021  09:59 AM           194,591 02-Conditional-Filtering.ipynb
08/05/2021  09:59 AM           196,385 03-Useful-Methods.ipynb
08/05/2021  09:59 AM            64,227 04-Missing-Data.ipynb
08/05/2021  09:59 AM           219,627 05-Groupby-Operations-and-MultiIndex.ipynb
08/05/2021  09:59 AM            62,966 06-Combining-DataFrames.ipynb
08/05/2021  09:59 AM            32,972 07-Text-Methods.ipynb
08/05/2021  09:59 AM            93,392 08-Time-Methods.ipynb
08/05/2021  09:59 AM            65,234 09-Inputs-and-Outputs.i

Find the 'csv' file you're looking for -
assign it the variable, using 'pd' as our instance of Pandas, as well as calling the function 'read_csv' and passing in the file path.

In [78]:
df = pd.read_csv('tips.csv')

In [79]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

Retrieve the column names (across the x-axis) -

In [80]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

Retrieve index information (down the y-axis, will show full text index if it exists) -

In [81]:
df.index

RangeIndex(start=0, stop=244, step=1)

Retrieve the top (supplied number) rows -

In [82]:
df.head(3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458

Retrieve the bottom (supplied number) rows -

In [83]:
df.tail(4)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

Again gives information on entire dataframe -

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


Retrieve length of the dataframe -

In [85]:
len(df)

244

Gives statistical information from each column -

In [86]:
df.describe()

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

As above, transposed to read across - 

In [87]:
df.describe().transpose()

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [88]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

#### Grab a Single Column

In [89]:
df['total_bill']

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [90]:
type(df['total_bill'])

pandas.core.series.Series

#### Grab Multiple Columns

Note we're passing in a python list of column names, hence the double brackets -


In [91]:
df[['total_bill','tip']]

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

Perform math how we would in a Numpy arrray, then assign it to a new column by referencing it as though it already exists -

In [92]:
df['tip_percentage'] = 100* df['tip'] / df['total_bill']

To view an example -

In [93]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

If we call a column that already exists, the data will be written over -

In [94]:
df['price_per_person'] = np.round(df['total_bill'] / df['size'], 2)

#### Adjust exisisting column data

Use 'np.round' on the calculation as well as a number of decimal points to round the answer. Either above, or below, retrieve the figures, round them and replace -

In [95]:
df['tip_percentage'] = np.round(df['tip_percentage'],2)

In [96]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959            5.94  
1      Douglas Tucker  4478071379779230    Sun4608           16.05  
2      Travis Walters  6011812112971322    Sun4458           16.66  
3    Nathaniel Harris  4676137647685994    Sun5260           13.98  
4        Tonya Carter  4832732618637221    Sun2251           14.68

In [97]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Remove Columns

Ensure axis is provided, of the function will not recognise -

Passing the dataframe back into itself with the dropped column makes the change stick

Axis 0 refers to the rows, where 1 is the columns.

In [98]:
df = df.drop("tip_percentage",axis=1)

In [99]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [100]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [101]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

It is important that data is indexed with a unique value for each column. As we have not defined an index, one is automatically created with sequential numbers, thus ensuring each row has a unique value -

In [102]:
df.index

RangeIndex(start=0, stop=244, step=1)

Using 'set_index' we can use a column with unique individual values to set as the array's index ('payment_id' in this case) -

The index will now move to the left column, and is technically now not a column, but the index.

In [103]:
df.set_index('Payment ID')

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

But the changes are not permanent....

In [104]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

So we can reassign it to the df variable name to make the changes permanent -

In [105]:
df = df.set_index('Payment ID')

In [106]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

We can also reset the index to the range number values (0 - 244) -

In [107]:
df = df.reset_index()

In [108]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

### ROWS

Let's now explore these same concepts but with Rows.

In [109]:
df.head()

Payment ID  total_bill   tip     sex smoker  day    time  size  \
0    Sun2959       16.99  1.01  Female     No  Sun  Dinner     2   
1    Sun4608       10.34  1.66    Male     No  Sun  Dinner     3   
2    Sun4458       21.01  3.50    Male     No  Sun  Dinner     3   
3    Sun5260       23.68  3.31    Male     No  Sun  Dinner     2   
4    Sun2251       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number  
0              8.49  Christy Cunningham  3560325168603410  
1              3.45      Douglas Tucker  4478071379779230  
2              7.00      Travis Walters  6011812112971322  
3             11.84    Nathaniel Harris  4676137647685994  
4              6.15        Tonya Carter  4832732618637221

In [110]:
df = df.set_index('Payment ID')

In [111]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Grab a Single Row

Use 'iloc' to retrieve a row on its integer-based location -


In [112]:
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

Or 'loc' to retrieve from a row's name-based location -


In [113]:
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

#### Grab Multiple Rows

Use Python's slice notation -

In [114]:
df.iloc[0:4]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

Or comma-seperated name locations -

In [115]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [116]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

As with 'drop' for a column, we must label the axis (0) for the rows -

In [117]:
df.drop('Sun2959',axis=0).head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

...again we must reassign the array to the variable to make the changes permanent

Error if you have a named index and pass in an index number -


In [118]:
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

Ensure the data is arranged so the array can interpret it!

In [121]:
one_row = df.iloc[0]

In [122]:
one_row

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [123]:
type(one_row)

pandas.core.series.Series

In [124]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  
Payment ID                                                       
Sat2657                 9.68    Michael Avila  5296068606052842  
Sat1766                13.59   Monica Sanders  3506806155565404  
Sat3880                11.34       Keith Wong  6011891618747196  
Sat17                   8.91     Dennis Dixon     4375220550950  
Thur672                 9.39  Michelle Hardin  3511451626698139

In [125]:
df.append(one_row).tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  
Sun2959                 8.49  Christy Cunningham  3560325168603410

--------